src/query-builder/SelectQueryBuilder.ts

```diff
@@ -1531,13 +1531,22 @@ export class SelectQueryBuilder<Entity> extends QueryBuilder<Entity> implements
         }
 
         if (this.connection.driver instanceof SqlServerDriver) {
+            let segment = "";
+
+            if ((limit || offset) && Object.keys(this.expressionMap.allOrderBys).length === 0) {
+                // MSSQL requires an ORDER BY if you use limits
+                this.addOrderBy(`${this.alias}.${this.expressionMap.mainAlias!.metadata.primaryColumns[0].propertyPath}`);
+                segment = this.createOrderByExpression();
```

Comment:
this confuses code order a bit, maybe better to move this logic on top before query building, something like if its sql server and no order buys but limit or offset presents we addOrderBy.